src/software/geom/spline.cpp

```diff
@@ -0,0 +1,92 @@
+#include "software/geom/spline.h"
+
+Spline::Spline(const std::vector<Point>& points) : knots(points)
+{
+    initLinearSegments(points);
+}
+
+Spline::Spline(const std::initializer_list<Point>& points) : knots(points)
+{
+    initLinearSegments(points);
+}
+
+Point Spline::valueAt(double val) const
+{
+    if (val < 0.0 || val > 1.0)
+    {
+        std::stringstream ss;
+        ss << "Tried to evaluate spline at " << val
+           << ", which is outside of domain of the spline: [0,1]";
+        throw std::invalid_argument(ss.str());
+    }
+
+    Point retval;
+
+    if (segments.empty())
+    {
+        retval = knots.front();
+    }
+    else
+    {
+        // Note: this could be more performant with binary search
```

Comment:
Maybe add an issue # for this.

Reply:
hm we don't know if it's a problem, so it's not actually something we need to fix atm